packages/core/src/external/carequality/ihe-gateway-v2/inbound/principal-and-delegates-cache.ts

```diff
@@ -0,0 +1,74 @@
+import { errorToString } from "@metriport/shared";
+import { out } from "../../../../util/log";
+import { capture } from "../../../../util/notifications";
+import { getPrincipalAndDelegatesMap } from "./principal-and-delegates";
+
+/**
+ * Cache for principal and delegates map to avoid repeated S3 calls in Lambda.
+ * This data is only updated during CQ directory rebuilds, so it's safe to cache
+ * at the Lambda level and reuse across invocations.
+ */
+class PrincipalAndDelegatesCache {
+  private static instance: PrincipalAndDelegatesCache;
+  private cache: Map<string, string[]> | undefined;
+  private isLoading = false;
+  private loadPromise: Promise<Map<string, string[]>> | undefined;
+
+  static getInstance(): PrincipalAndDelegatesCache {
+    if (!PrincipalAndDelegatesCache.instance) {
+      PrincipalAndDelegatesCache.instance = new PrincipalAndDelegatesCache();
+    }
+    return PrincipalAndDelegatesCache.instance;
+  }
+
+  /**
+   * Get the principal and delegates map from cache or load it if not cached.
+   * This method is safe to call multiple times - it will only load once.
+   */
+  async getMap(): Promise<Map<string, string[]>> {
+    if (this.cache) {
+      return this.cache;
+    }
+
+    if (this.isLoading && this.loadPromise) {
+      return this.loadPromise;
+    }
+
+    this.isLoading = true;
+    this.loadPromise = this.loadMapFromS3();
+
+    try {
+      this.cache = await this.loadPromise;
+      return this.cache;
+    } catch (error) {
+      this.isLoading = false;
+      this.loadPromise = undefined;
+      throw error;
+    }
```

Comment:
This looks overly complicated (returning `loadPromise`, etc.).

`getMap` should just
- if cached, return it
- otherwise, get it, update the cache, return the cache